src/software/gui/full_system/widgets/ai_control.cpp

```diff
@@ -163,3 +166,96 @@ void setupPlayOverrideComboBox(Ui::AutogeneratedFullSystemMainWidget *widget,
                      on_play_changed);
 }
+
+void setupChannelSpinBox(Ui::AutogeneratedFullSystemMainWidget *widget,
+                         std::shared_ptr<ThunderbotsConfig> config)
+{
+    auto parameter = config->getMutableNetworkConfig()->mutableChannel();
+    auto spin_box  = widget->channel_spin_box;
+
+    spin_box->setValue(parameter->value());
+    spin_box->setMinimum(0);
+    spin_box->setMaximum(MAX_MULTICAST_CHANNELS);
+
+    auto on_channel_changed = [parameter](int new_value) {
+        parameter->setValue(new_value);
+    };
+
+    // QSpinBox has 2 "valueChanged" signals that each provide different info (string vs
+    // int), so we need to static_cast to specify the integer version
+    QWidget::connect(spin_box,
+                     static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged),
+                     on_channel_changed);
+
+    auto on_parameter_value_changed = [spin_box](int new_value) {
+        // We block signals while setting the value of the spinbox so that we don't
+        // trigger the `on_spinbox_value_changed` function, which would set the
+        // parameter value again and deadlock on the parameter's internal mutex
+        spin_box->blockSignals(true);
+        spin_box->setValue(new_value);
+        spin_box->blockSignals(false);
+    };
+    parameter->registerCallbackFunction(on_parameter_value_changed);
+}
+
+void setupFriendlyGoalieIDSpinBox(Ui::AutogeneratedFullSystemMainWidget *widget,
+                                  std::shared_ptr<ThunderbotsConfig> config)
+{
+    auto parameter = config->getMutableSensorFusionConfig()->mutableFriendlyGoalieId();
+    auto spin_box  = widget->friendly_goalie_id_spin_box;
+
+    spin_box->setValue(parameter->value());
+    spin_box->setMinimum(0);
+    spin_box->setMaximum(std::numeric_limits<uint8_t>::max());
+
+    auto on_channel_changed = [parameter](int new_value) {
+        parameter->setValue(new_value);
+    };
+
+    // QSpinBox has 2 "valueChanged" signals that each provide different info (string vs
+    // int), so we need to static_cast to specify the integer version
+    QWidget::connect(spin_box,
+                     static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged),
+                     on_channel_changed);
+
+    auto on_parameter_value_changed = [spin_box](int new_value) {
+        // We block signals while setting the value of the spinbox so that we don't
+        // trigger the `on_spinbox_value_changed` function, which would set the
+        // parameter value again and deadlock on the parameter's internal mutex
+        spin_box->blockSignals(true);
+        spin_box->setValue(new_value);
+        spin_box->blockSignals(false);
+    };
+    parameter->registerCallbackFunction(on_parameter_value_changed);
+}
+
+void setupEnemyGoalieIDSpinBox(Ui::AutogeneratedFullSystemMainWidget *widget,
+                               std::shared_ptr<ThunderbotsConfig> config)
```

Comment:
Any reason we pass the whole ThunderbotsConfig vs just the mutableEnemyGoalieID() that's used?

Reply:
Not really. Updated all functions to just take the parameters they need.